File: TranslatorStudioClassLibrary/TranslatorStudioClassLibrary/Class/TranslationData.cs
namespace TranslatorStudioClassLibrary.Class
{
    using System.Collections.Generic;
    using System.Linq;
    using Exception;
    using Interface;
    using Utilities;

    /// <summary>
    /// Class that contains the properties and method relevant for Translation Data.
    /// Implements Translation Data Interface.
    /// </summary>
    public class TranslationData : ITranslationData //, IEquatable<TranslationData>
    {
        #region Properties

        #region Project Data

        /// <summary>
        /// Private property that constructs sub translation data;
        /// </summary>
        private ISubTranslationDataFactory _subTranslationDataFactory;

        /// <summary>
        /// Private property that contains Project Data.
        /// </summary>
        private IProjectData _data { get; set; }
        /// <summary>
        /// Private property that contains Sub Translation Data.
        /// </summary>
        private ISubTranslationData _subData { get; set; }
        /// <summary>
        /// Private property that contains Auto Translation Data.
        /// </summary>
        private ISubTranslationData _autoData { get; set; }

        /// <summary>
        /// Private property that contains Auto Translation Mode flag.
        /// </summary>
        private bool _autoMode { get; set; } = false;
        /// <summary>
        /// Determines whether Default Translation Mode is turned on or not.
        /// </summary>
        public bool DefaultTranslationMode { get; set; } = true;
        /// <summary>
        /// Determines whether Auto Translation Mode is turned on or not.
        /// </summary>
        public bool AutoTranslationMode { get => _autoMode; }
        /// <summary>
        /// Determines whether changed have been made to Translation Data.
        /// </summary>
        public bool DataChanged { get; set; }

        /// <summary>
        /// The name of the project.
        /// </summary>
        public string ProjectName
        {
            get => _data.ProjectName;
            set => _data.ProjectName = SetPropertyValue(value);
        }
        /// <summary>
        /// The url to the project source.
        /// </summary>
        public string SourceLink
        {
            get => _data.ProjectName;
            set => _data.ProjectName = SetPropertyValue(value);
        }
        /// <summary>
        /// Contains all lines in the translation.
        /// </summary>
        public IList<IProjectLine> ProjectLines { get => _data.ProjectLines; set => _data.ProjectLines = value; }

        #endregion

        #region Project Controls

        /// <summary>
        /// The current project line in the translation.
        /// </summary>
        public IProjectLine CurrentLine
        {
            get
            {
                var index = DefaultTranslationMode ?
                   (AutoTranslationMode ? _autoData.CurrentReference : CurrentIndex)
                   : _subData.CurrentReference;

                return ProjectLines[index];
            }
            set
            {
                var index = DefaultTranslationMode ?
                    (AutoTranslationMode ? _autoData.CurrentReference : CurrentIndex)
                    : _subData.CurrentReference;

                ProjectLines[index] = SetPropertyValue(value);
            }
        }

        /// <summary>
        /// The current raw line in the translation.
        /// </summary>
        public string CurrentRaw { get => CurrentLine.Raw; set => CurrentLine.Raw = SetPropertyValue(value); }
        /// <summary>
        /// The current translated line in the translation.
        /// </summary>
        public string CurrentTranslation { get => CurrentLine.Translation; set => CurrentLine.Translation = SetPropertyValue(value); }
        /// <summary>
        /// The comment of the current line in the translation.
        /// </summary>
        public string CurrentComment { get => CurrentLine.Comment; set => CurrentLine.Comment = SetPropertyValue(value); }
        /// <summary>
        /// The completion status of the current line in the translation.
        /// </summary>
        public bool CurrentCompletion { get => CurrentLine.Completed; set => CurrentLine.Completed = SetPropertyValue(value); }
        /// <summary>
        /// The marked status of the current line in the translation.
        /// </summary>
        public bool CurrentMarked { get => CurrentLine.Marked; set => CurrentLine.Marked = SetPropertyValue(value); }

        /// <summary>
        /// Private property that contains the current index.
        /// </summary>
        private int _index = 0;

        /// <summary>
        /// The current index in the translation.
        /// </summary>
        public int CurrentIndex
        {
            get
            {
                if (DefaultTranslationMode)
                    return AutoTranslationMode ? _autoData.CurrentIndex: _index;
                else
                    return _subData.CurrentIndex;
            }
            set
            {
                if (DefaultTranslationMode)
                    if (AutoTranslationMode)
                        _autoData.CurrentIndex = value;
                    else
                        _index = value;
                else
                    _subData.CurrentIndex = value;

            }
        }
        /// <summary>
        /// The max index in the translation.
        /// </summary>
        public int MaxIndex
        {
            get
            {
                if (DefaultTranslationMode)
                    return AutoTranslationMode ? _autoData.MaxIndex : ProjectLines.Count - 1;
                else
                    return _subData.MaxIndex;
            }
        }

        /// <summary>
        /// The number of lines in the translation.
        /// </summary>
        public int NumberOfLines => AutoTranslationMode ? _autoData.NumberOfLines : ProjectLines.Count;
        /// <summary>
        ///The number of completed lines in the translation.
        /// </summary>
        public int NumberOfCompletedLines
        {
            get
            {
                return ProjectLines
                    .Where(x => x.Completed == true && (AutoTranslationMode ? x.Raw.Any() : true))
                    .Select(x => x).ToList().Count();
            }
        }

        #endregion

        #endregion


        #region Property Changed
        
        private T SetPropertyValue<T>(T value)
        {
            DataChanged = true;
            return value;
        }

        #endregion


        #region Constructors

        /// <summary>
        /// Creates empty Translation Data.
        /// </summary>
        /// <param name="subTranslationDataFactory">Sub Translation Data Factory used to construct Sub Translation Data.</param>
        public TranslationData(ISubTranslationDataFactory subTranslationDataFactory)
        {
            _subTranslationDataFactory = subTranslationDataFactory ?? throw new System.ArgumentNullException(nameof(subTranslationDataFactory));
        }

        /// <summary>
        /// Creates Translation Data based on passed Project Data.
        /// </summary>
        /// <param name="projectData">Object that implements Project Data Interface.</param>
        /// <param name="subTranslationDataFactory"></param>
        public TranslationData(IProjectData projectData, ISubTranslationDataFactory subTranslationDataFactory) : this(subTranslationDataFactory)
        {
            _data = projectData ?? throw new System.ArgumentNullException(nameof(projectData));
            CurrentIndex = 0;
        }

        #endregion


        #region Methods

        /// <summary>
        /// Increments the current index.
        /// </summary>
        public void IncrementCurrentLine()
        {
            if (CurrentIndex != MaxIndex)
            {
                CurrentIndex++;
                if (_autoMode)
                {
                    var testIndex = _autoData.CurrentReference != 0 ? _autoData.CurrentReference - 1 : 0;
                    if (string.IsNullOrWhiteSpace(ProjectLines[testIndex].Raw))
                        ProjectLines[testIndex].Completed = true;
                }
            }
        }
        /// <summary>
        /// Decrements the current index.
        /// </summary>
        public void DecrementCurrentLine()
        {
            if (CurrentIndex != 0)
            {
                CurrentIndex--;
                if (_autoMode)
                {
                    var testIndex = _autoData.CurrentReference != 0 ? _autoData.CurrentReference - 1 : 0;
                    if (string.IsNullOrWhiteSpace(ProjectLines[testIndex].Raw))
                        ProjectLines[testIndex].Completed = true;
                }
            }
        }

        /// <summary>
        /// Inserts specified raw line to project data at index.
        /// </summary>
        /// <param name="index">Index at which to insert value (null will insert value at end).</param>
        /// <param name="rawValue">Value of raw line to insert.</param>
        public void InsertLine(int? index, string rawValue)
        {
            int insertionIndex = index ?? NumberOfLines;

            rawValue = rawValue ?? "";

            var newLine = new ProjectLine
            {
                Raw = rawValue,
                Translation = "",
                Completed = false,
                Marked = false
            };

            ProjectLines.Insert(insertionIndex, newLine);
        }
        /// <summary>
        /// Removes line from project data at index.
        /// </summary>
        /// <param name="index">Index at which to remove value (null will remove value at end).</param>
        /// <exception cref="RemovalOfLastLineException">Thrown when trying to remove last line of translation project.</exception>
        public void RemoveLine(int? index)
        {
            if (MaxIndex == 0)
                throw ExceptionHelper.NewRemovalOfLastLineException;
            int insertionIndex = index ?? MaxIndex;

            ProjectLines.RemoveAt(insertionIndex);
        }

        /// <summary>
        /// Gets JSON string of the project data used to save project.
        /// </summary>
        /// <returns>JSON string of project data.</returns>
        public string GetProjectSaveString()
        {
            return _data.GetSaveString();
        }

        /// <summary>
        /// Toggles auto translation mode. (Removes empty lines.)
        /// </summary>
        /// <param name="autoOn">Whether or not auto mode should be turned on or not.</param>
        /// <returns>Number of lines that are not empty.</returns>
        public int ToggleAutoMode(bool autoOn)
        {
            if (autoOn)
            {
                var rawLines = ProjectLines.Select(x => x.Raw).ToList();
                var nonEmptyRaw = rawLines.Select(x => x.IsNotEmpty()).ToList();
                _autoData = _subTranslationDataFactory.GetSubData(nonEmptyRaw);

                _autoMode = true;
                CurrentIndex = 0;
            }
            else
                _autoMode = false;

            return NumberOfLines;
        }
        /// <summary>
        /// Initiates default translation mode.
        /// </summary>
        public void StartDefaultMode()
        {
            DefaultTranslationMode = true;
            CurrentIndex = 0;
        }
        /// <summary>
        /// Initiates marked only translation mode. (Shows all marked lines only).
        /// </summary>
        /// <returns>Number of lines that are marked.</returns>
        public int StartMarkedOnlyMode()
        {
            List<bool> conditionList;
            var rawLines = ProjectLines.Select(x => x.Raw).ToList();
            var markedLines = ProjectLines.Select(x => x.Marked).ToList();
            if (_autoMode)
                conditionList = rawLines.Select((v, i) => new { v, i }).Select(x => markedLines[x.i] && x.v.IsNotEmpty()).ToList();
            else
                conditionList = markedLines;

            _subData = _subTranslationDataFactory.GetSubData(conditionList);
            DefaultTranslationMode = false;
            CurrentIndex = 0;
            return _subData.NumberOfLines;
        }
        /// <summary>
        /// Initates incomplete only translation mode. (Shows all incomplete lines only.)
        /// </summary>
        /// <returns>Number of lines that are incomplete.</returns>
        public int StartIncompleteOnlyMode()
        {
            List<bool> conditionList;
            var rawLines = ProjectLines.Select(x => x.Raw).ToList();
            var completedLines = ProjectLines.Select(x => x.Completed).ToList();
            if (_autoMode)
                conditionList = rawLines.Select((v, i) => new { v, i }).Select(x => !completedLines[x.i] && x.v.IsNotEmpty()).ToList();
            else
                conditionList = completedLines.Select(x => !x).ToList();

            _subData = _subTranslationDataFactory.GetSubData(conditionList);
            DefaultTranslationMode = false;
            CurrentIndex = 0;
            return _subData.NumberOfLines;
        }
        /// <summary>
        /// Initiates complete only translation mode. (Shows all complete lines only.)
        /// </summary>
        /// <returns>Number of lines that are complete.</returns>
        public int StartCompleteOnlyMode()
        {
            List<bool> conditionList;
            var rawLines = ProjectLines.Select(x => x.Raw).ToList();
            var completedLines = ProjectLines.Select(x => x.Completed).ToList();
            if (_autoMode)
                conditionList = rawLines.Select((v, i) => new { v, i }).Select(x => completedLines[x.i] && x.v.IsNotEmpty()).ToList();
            else
                conditionList = completedLines;

            _subData = _subTranslationDataFactory.GetSubData(conditionList);
            DefaultTranslationMode = false;
            CurrentIndex = 0;
            return _subData.NumberOfLines;
        }

        /// <summary>
        /// Returns the project data used in the translation project.
        /// </summary>
        /// <returns>Object that implements Project Data Interface.</returns>
        public IProjectData GetProjectData()
        {
            return _data;
        }

        #endregion

        //public bool Equals(TranslationData other)
        //{
        //    return other != null &&
        //           EqualityComparer<ISubTranslationDataFactory>.Default.Equals(_subTranslationDataFactory, other._subTranslationDataFactory) &&
        //           EqualityComparer<IProjectData>.Default.Equals(_data, other._data) &&
        //           EqualityComparer<ISubTranslationData>.Default.Equals(_subData, other._subData) &&
        //           EqualityComparer<ISubTranslationData>.Default.Equals(_autoData, other._autoData) &&
        //           _autoMode == other._autoMode &&
        //           DefaultTranslationMode == other.DefaultTranslationMode &&
        //           AutoTranslationMode == other.AutoTranslationMode &&
        //           DataChanged == other.DataChanged &&
        //           ProjectName == other.ProjectName &&
        //           EqualityComparer<List<IProjectLine>>.Default.Equals(ProjectLines, other.ProjectLines) &&
        //           EqualityComparer<IProjectLine>.Default.Equals(CurrentLine, other.CurrentLine) &&
        //           CurrentRaw == other.CurrentRaw &&
        //           CurrentTranslation == other.CurrentTranslation &&
        //           CurrentCompletion == other.CurrentCompletion &&
        //           CurrentMarked == other.CurrentMarked &&
        //           _index == other._index &&
        //           CurrentIndex == other.CurrentIndex &&
        //           MaxIndex == other.MaxIndex &&
        //           NumberOfLines == other.NumberOfLines &&
        //           NumberOfCompletedLines == other.NumberOfCompletedLines;
        //}

        //public override int GetHashCode()
        //{
        //    var hashCode = 471267684;
        //    hashCode = hashCode * -1521134295 + EqualityComparer<ISubTranslationDataFactory>.Default.GetHashCode(_subTranslationDataFactory);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<IProjectData>.Default.GetHashCode(_data);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<ISubTranslationData>.Default.GetHashCode(_subData);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<ISubTranslationData>.Default.GetHashCode(_autoData);
        //    hashCode = hashCode * -1521134295 + _autoMode.GetHashCode();
        //    hashCode = hashCode * -1521134295 + DefaultTranslationMode.GetHashCode();
        //    hashCode = hashCode * -1521134295 + AutoTranslationMode.GetHashCode();
        //    hashCode = hashCode * -1521134295 + DataChanged.GetHashCode();
        //    hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(ProjectName);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<List<IProjectLine>>.Default.GetHashCode(ProjectLines);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<IProjectLine>.Default.GetHashCode(CurrentLine);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(CurrentRaw);
        //    hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(CurrentTranslation);
        //    hashCode = hashCode * -1521134295 + CurrentCompletion.GetHashCode();
        //    hashCode = hashCode * -1521134295 + CurrentMarked.GetHashCode();
        //    hashCode = hashCode * -1521134295 + _index.GetHashCode();
        //    hashCode = hashCode * -1521134295 + CurrentIndex.GetHashCode();
        //    hashCode = hashCode * -1521134295 + MaxIndex.GetHashCode();
        //    hashCode = hashCode * -1521134295 + NumberOfLines.GetHashCode();
        //    hashCode = hashCode * -1521134295 + NumberOfCompletedLines.GetHashCode();
        //    return hashCode;
        //}
    }
}

File: OnlineTranslatorStudio/OnlineTranslatorStudio/Javascript/TranslationData.ts
//import proj = require("ProjectData");
//import * as ko from "knockout";

//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes

interface ITranslationData {
    DefaultTranslationMode: KnockoutObservable<boolean>;

    // Properties
    ProjectName: KnockoutObservable<string>;
    ProjectLines: KnockoutObservableArray<KnockoutObservable<IProjectLineViewModel>>;
    RawReadOnly: KnockoutObservable<boolean>;

    // Controls
    CurrentIndex: KnockoutObservable<number>;
    MaxIndex: KnockoutComputed<number>;
    CurrentLineNumber: KnockoutComputed<number>;
    MaxLineNumber: KnockoutComputed<number>;

    CurrentLine: KnockoutObservable<IProjectLineViewModel>;
    CurrentRaw: KnockoutObservable<string>;
    CurrentTranslation: KnockoutObservable<string>;
    CurrentMarked: KnockoutObservable<boolean>;
    CurrentCompletion: KnockoutObservable<boolean>;

    // Computed
    NumberOfLines: KnockoutComputed<number>;
    NumberOfCompletedLines: KnockoutComputed<number>;

    CurrentProgress: KnockoutComputed<number>;

    // Methods
    IncrementIndex: () => void;
    DecrementIndex: () => void;

    InsertLine: (index: number) => void;
    RemoveLine: (index: number) => void;

    StartDefaultMode: () => void;
    StartMarkedOnlyMode: () => void;
    StartIncompleteOnlyMode: () => void;
    StartCompleteOnlyMode: () => void;


    ToggleRawReadOnly: (value: boolean) => void;

    GetSaveString: () => string;
}

class TranslationData implements ITranslationData {
    private _projectData: KnockoutObservable<IProjectViewModel>;
    private _subData: ISubTranslationData;
    private _index: KnockoutObservable<number>;
    private _maxIndex: KnockoutComputed<number>;

    DefaultTranslationMode: KnockoutObservable<boolean>;
    RawReadOnly: KnockoutObservable<boolean>;

    NumberOfLines: KnockoutComputed<number>;
    NumberOfCompletedLines: KnockoutComputed<number>;

    CurrentProgress: KnockoutComputed<number>;

    // Constructor
    constructor(projectData: IProjectViewModel) {
        if (projectData == undefined) throw Error('Project Data is undefined');
        this._projectData = ko.observable(projectData);
        
        this._index = ko.observable(0);

        this.DefaultTranslationMode = ko.observable(true);
        this.RawReadOnly = ko.observable(true);

        this._maxIndex = ko.computed({
            owner: this,
            read: () => {
                return this.ProjectLines().length - 1;
            }
        });
        this.NumberOfLines = ko.computed({
            owner: this,
            read: () => {
                return this.ProjectLines().length;
            }
        });
        this.NumberOfCompletedLines = ko.computed({
            owner: this,
            read: () => {
                var result = 0;
                for (var i = 0; i < this.ProjectLines().length; i++) {
                    if (this.ProjectLines()[i]().Completed() == true) {
                        result++;
                    }
                }
                return result;
            }
        });

        this.CurrentProgress = ko.computed({
            owner: this,
            read: () => {
                return +(((this.NumberOfCompletedLines() / this.NumberOfLines()) * 100).toFixed(2));
            }
        });

        ko.applyBindings(this);
    }

    // Properties
    get ProjectName(): KnockoutObservable<string> {
        return this._projectData().projectName;
    }
    set ProjectName(value: KnockoutObservable<string>) {
        this._projectData().projectName(value());
    }

    get ProjectLines(): KnockoutObservableArray<KnockoutObservable<IProjectLineViewModel>> {
        return this._projectData().projectLines;
    }

    set ProjectLines(value: KnockoutObservableArray<KnockoutObservable<IProjectLineViewModel>>) {
        this._projectData().projectLines(value());
    }

    //Controls
    get CurrentIndex(): KnockoutObservable<number> {
        if (this.DefaultTranslationMode()) {
            return this._index;
        }
        else {
            return this._subData.CurrentIndex;
        }
    }
    set CurrentIndex(value: KnockoutObservable<number>) {
        if (this.DefaultTranslationMode()) {
            this._index(value());
        }
        else {
            this._subData.CurrentIndex(value());
        }
    }

    get MaxIndex(): KnockoutComputed<number> {
        if (this.DefaultTranslationMode()) {
            return this._maxIndex;
        }
        else {
            return this._subData.MaxIndex;
        }
    }

    get CurrentLineNumber(): KnockoutComputed<number> {
        return ko.pureComputed({
            owner: this,
            read: () => {
                return this.CurrentIndex() + 1;
            },
            write: (value) => {
                this.CurrentIndex(value - 1);
            }
        });
    }

    get MaxLineNumber(): KnockoutComputed<number> {
        return ko.computed({
            owner: this,
            read: () => {
                return this.MaxIndex() + 1;
            }
        });
    }

    get CurrentLine(): KnockoutObservable<IProjectLineViewModel> {
        var index: number;
        if (this.DefaultTranslationMode()) {
            index = this.CurrentIndex();
        }
        else {
            index = this._subData.CurrentReference();
        }
        return this.ProjectLines()[index];
    }
    set CurrentLine(value: KnockoutObservable<IProjectLineViewModel>) {
        var index: number;
        if (this.DefaultTranslationMode()) {
            index = this.CurrentIndex();
        }
        else {
            index = this._subData.CurrentReference();
        }
        this.ProjectLines()[index](value());
    }

    get CurrentRaw(): KnockoutObservable<string> {
        return this.CurrentLine().Raw;
    }
    set CurrentRaw(value: KnockoutObservable<string>) {
        this.CurrentLine().Raw(value());
    }

    get CurrentTranslation(): KnockoutObservable<string> {
        return this.CurrentLine().Translation;
    }
    set CurrentTranslation(value: KnockoutObservable<string>) {
        this.CurrentLine().Translation(value());
    }

    get CurrentCompletion(): KnockoutObservable<boolean> {
        return this.CurrentLine().Completed;
    }
    set CurrentCompletion(value: KnockoutObservable<boolean>) {
        this.CurrentLine().Completed(value());
    }

    get CurrentMarked(): KnockoutObservable<boolean> {
        return this.CurrentLine().Marked;
    }
    set CurrentMarked(value: KnockoutObservable<boolean>) {
        this.CurrentLine().Marked(value());
    }

    // Methods
    IncrementIndex(): void {
        if (this.CurrentIndex() < this.MaxIndex()) {
            this.CurrentIndex(this.CurrentIndex() + 1);
        }
    }

    DecrementIndex(): void {
        if (this.CurrentIndex() > 0) {
            this.CurrentIndex(this.CurrentIndex() - 1);
        }
    }


    InsertLine(index: number): void {
        if (index == undefined) index = this.NumberOfLines();
        var newData: IProjectLine = { Raw: "", Translation: "", Completed: false, Marked: false };
        var newLine: KnockoutObservable<IProjectLineViewModel> = ko.observable(new ProjectLineViewModel(newData));
        this.ProjectLines.splice(index, 0, newLine);
    }

    RemoveLine(index: number): void {
        if (index == undefined) index = this.NumberOfLines();
        this.ProjectLines.splice(index, 1);
    }


    StartDefaultMode(): void {
        this.DefaultTranslationMode(true);
        this.CurrentIndex(0);
    }

    StartMarkedOnlyMode(): void {
        if (this.DefaultTranslationMode() == false) {
            this.StartDefaultMode();
        }

        var conditionList: boolean[] = [];

        for (var i = 0; i < this.ProjectLines().length; i++) {
            conditionList.push(this.ProjectLines()[i]().Marked());
        }

        try {
            this._subData = new SubTranslationData(conditionList);

            this.DefaultTranslationMode(false);
            this.CurrentIndex(0);

        } catch (e) {
            if ((<Error>e).message == "Condition List returned no results. Cannot construct Sub Translation Data") {
                alert("No marked lines. Returning to Default Mode...");
            }
            else {
                alert("Something went wrong. Returning to Default Mode...");
            }
            this.StartDefaultMode();
            throw Error(e);
        }
    }

    StartIncompleteOnlyMode(): void {
        if (this.DefaultTranslationMode() == false) {
            this.StartDefaultMode();
        }

        var conditionList: boolean[] = [];

        for (var i = 0; i < this.ProjectLines().length; i++) {
            conditionList.push(!this.ProjectLines()[i]().Completed());
        }

        try {
            this._subData = new SubTranslationData(conditionList);

            this.DefaultTranslationMode(false);
            this.CurrentIndex(0);

        } catch (e) {
            if ((<Error>e).message == "Condition List returned no results. Cannot construct Sub Translation Data") {
                alert("No incomplete lines. Returning to Default Mode...");
            }
            else {
                alert("Something went wrong. Returning to Default Mode...");
            }
            this.StartDefaultMode();
            throw Error(e);
        }
    }

    StartCompleteOnlyMode(): void {
        if (this.DefaultTranslationMode() == false) {
            this.StartDefaultMode();
        }

        var conditionList: boolean[] = [];

        for (var i = 0; i < this.ProjectLines().length; i++) {
            conditionList.push(this.ProjectLines()[i]().Completed());
        }

        try {
            this._subData = new SubTranslationData(conditionList);

            this.DefaultTranslationMode(false);
            this.CurrentIndex(0);

        } catch (e) {
            if ((<Error>e).message == "Condition List returned no results. Cannot construct Sub Translation Data") {
                alert("No complete lines. Returning to Default Mode...");
            }
            else {
                alert("Something went wrong. Returning to Default Mode...");
            }
            this.StartDefaultMode();
            throw Error(e);
        }
    }


    ToggleRawReadOnly(value?: boolean): void {
        if (value == undefined) {
            this.RawReadOnly(!this.RawReadOnly());
        }
        else {
            this.RawReadOnly(value);
        }
    }

    GetSaveString(): string {
        var projectData: IProjectData = new ProjectData(this._projectData());
        return JSON.stringify({ data: projectData });
    }
}